File: internal/tools/read.go
package tools

import (
	"context"
	"os"
	"strings"

	"minicode/internal/domain"
)

// ReadTool reads file content.
type ReadTool struct{}

// NewReadTool creates a ReadTool.
func NewReadTool() *ReadTool { return &ReadTool{} }

func (t *ReadTool) Name() string        { return "Read" }
func (t *ReadTool) Description() string { return "Read the content of a file. Returns the file content. For large files, use offset and limit to read in chunks." }
func (t *ReadTool) RequiresPermission() bool { return false }
func (t *ReadTool) Requires() []ToolRequirement { return nil }

func (t *ReadTool) InputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path":   map[string]any{"type": "string"},
			"offset": map[string]any{"type": "integer"},
			"limit":  map[string]any{"type": "integer"},
		},
		"required": []string{"path"},
	}
}

func (t *ReadTool) Execute(ctx context.Context, args map[string]any, tc ToolContext) (domain.ToolResult, error) {
	path, _ := args["path"].(string)
	if path == "" {
		return domain.ToolResult{Output: "Error: path is required"}, nil
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return domain.ToolResult{Output: err.Error()}, nil
	}

	content := string(data)
	lines := strings.Split(content, "\n")

	start := 0
	if offset, ok := args["offset"].(float64); ok && offset > 0 {
		start = int(offset) - 1
	}

	end := len(lines)
	if limit, ok := args["limit"].(float64); ok {
		end = start + int(limit)
	}

	if start < 0 {
		start = 0
	}
	if start > len(lines) {
		start = len(lines)
	}
	if end < start {
		end = start
	}
	if end > len(lines) {
		end = len(lines)
	}

	result := strings.Join(lines[start:end], "\n")
	return domain.ToolResult{Output: result}, nil
}
func init() { Register(NewReadTool()) }

File: src/tools/builtin/read.ts
import fs from "fs/promises";
import path from "path";
import type { LLMMediaType } from "../../core/blocks.js";
import type {
  ToolDef,
  ToolRunResult,
  ToolExecutionContext,
} from "../registry.js";

const IMAGE_EXTENSIONS: Record<string, LLMMediaType> = {
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
};

/** Providers cap image payloads around 5 MB — refuse beyond that rather
 *  than bloat every subsequent request with data the API will reject. */
const MAX_IMAGE_BYTES = 5 * 1024 * 1024;

function imageMediaType(filePath: string): LLMMediaType | undefined {
  return IMAGE_EXTENSIONS[path.extname(filePath).toLowerCase()];
}

function formatBytes(bytes: number): string {
  if (bytes >= 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)}MB`;
  return `${Math.max(1, Math.round(bytes / 1024))}KB`;
}

async function readImage(
  filePath: string,
  mediaType: LLMMediaType,
  vision: boolean,
): Promise<ToolRunResult> {
  try {
    const stat = await fs.stat(filePath);
    const label = `[image: ${filePath} (${mediaType}, ${formatBytes(stat.size)})]`;

    if (!vision) {
      return {
        outcome: "success",
        result: `${label} Model does not support vision; image content not loaded.`,
      };
    }
    if (stat.size > MAX_IMAGE_BYTES) {
      return {
        outcome: "error",
        reason: `${filePath} is ${formatBytes(stat.size)} — images larger than ${formatBytes(MAX_IMAGE_BYTES)} are not supported.`,
      };
    }

    const base64 = (await fs.readFile(filePath)).toString("base64");
    return {
      outcome: "success",
      result: label,
      images: [{ mediaType, base64 }],
    };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return { outcome: "error", reason: msg };
  }
}

export const readTool: ToolDef = {
  name: "Read",
  description:
    "Read the contents of a file. Supports text files and images (png/jpg/jpeg/gif/webp — sent to the model when it supports vision). Defaults to first 2000 lines. Use offset/limit for large files.",
  readOnly: true,
  input_schema: {
    type: "object" as const,
    properties: {
      path: { type: "string", description: "Path to the file" },
      offset: {
        type: "number",
        description: "Line number to start from (1-indexed)",
      },
      limit: { type: "number", description: "Maximum number of lines to read" },
    },
    required: ["path"],
  },
  execute: async (
    args: Record<string, unknown>,
    context?: ToolExecutionContext,
  ): Promise<ToolRunResult> => {
    try {
      const path = args.path as string;
      const offset = args.offset as number | undefined;
      const limit = args.limit as number | undefined;

      const mediaType = imageMediaType(path);
      if (mediaType) {
        return readImage(
          path,
          mediaType,
          context?.config.model.supportsVision() ?? false,
        );
      }

      const content = await fs.readFile(path, "utf-8");
      const lines = content.split("\n");
      const start = (offset || 1) - 1;
      const end = limit ? start + limit : lines.length;
      const result = lines.slice(start, end).join("\n");
      return { outcome: "success", result: result };
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      return { outcome: "error", reason: msg };
    }
  },
};
